datafusion/datasource-parquet/src/metadata.rs

```diff
@@ -0,0 +1,564 @@
+}
+
+impl<'a> DFParquetMetadata<'a> {
+    pub fn new(store: &'a dyn ObjectStore, object_meta: &'a ObjectMeta) -> Self {
+        Self {
+            store,
+            object_meta,
+            metadata_size_hint: None,
+            decryption_properties: None,
+            file_metadata_cache: None,
+            coerce_int96: None,
+        }
+    }
+
+    /// set metadata size hint
+    pub fn with_metadata_size_hint(mut self, metadata_size_hint: Option<usize>) -> Self {
+        self.metadata_size_hint = metadata_size_hint;
+        self
+    }
+
+    /// set decryption properties
+    pub fn with_decryption_properties(
+        mut self,
+        decryption_properties: Option<&'a FileDecryptionProperties>,
+    ) -> Self {
+        self.decryption_properties = decryption_properties;
+        self
+    }
+
+    /// set file metadata cache
+    pub fn with_file_metadata_cache(
+        mut self,
+        file_metadata_cache: Option<Arc<dyn FileMetadataCache>>,
+    ) -> Self {
+        self.file_metadata_cache = file_metadata_cache;
+        self
+    }
+
+    /// Set timeunit to coerce INT96 timestamps to
+    pub fn with_coerce_int96(mut self, time_unit: Option<TimeUnit>) -> Self {
+        self.coerce_int96 = time_unit;
+        self
+    }
+
+    /// fetch parquet metadata
+    pub async fn fetch_metadata(&self) -> Result<Arc<ParquetMetaData>> {
+        let Self {
+            store,
+            object_meta,
+            metadata_size_hint,
+            decryption_properties,
+            file_metadata_cache,
+            coerce_int96: _,
+        } = self;
+
+        let fetch = ObjectStoreFetch::new(*store, object_meta);
+
+        // implementation to fetch parquet metadata
+        let cache_metadata =
+            !cfg!(feature = "parquet_encryption") || decryption_properties.is_none();
+
+        if cache_metadata {
+            if let Some(parquet_metadata) = file_metadata_cache
+                .as_ref()
+                .and_then(|file_metadata_cache| file_metadata_cache.get(object_meta))
+                .and_then(|file_metadata| {
+                    file_metadata
+                        .as_any()
+                        .downcast_ref::<CachedParquetMetaData>()
+                        .map(|cached_parquet_metadata| {
+                            Arc::clone(cached_parquet_metadata.parquet_metadata())
+                        })
+                })
+            {
+                return Ok(parquet_metadata);
+            }
+        }
+
+        let mut reader =
+            ParquetMetaDataReader::new().with_prefetch_hint(*metadata_size_hint);
+
+        #[cfg(feature = "parquet_encryption")]
+        if let Some(decryption_properties) = decryption_properties {
+            reader = reader.with_decryption_properties(Some(decryption_properties));
+        }
+
+        if cache_metadata && file_metadata_cache.is_some() {
+            // Need to retrieve the entire metadata for the caching to be effective.
```

Comment:
Does this mean that even if I have page indexes disabled if I use a metadata cache it will still retrieve (and decode?) them?